audio/iir_filter_node.rs

```diff
@@ -0,0 +1,177 @@
+
+const MAX_COEFFS: usize = 20;
+
+#[derive(Debug)]
+pub struct IIRFilterNodeOptions {
+    pub feedforward: Vec<f64>,
+    pub feedback: Vec<f64>,
+}
+
+#[derive(Clone)]
+struct IIRFilter {
+    feedforward: Arc<Vec<f64>>,
+    feedback: Arc<Vec<f64>>,
+    inputs: VecDeque<f64>,
+    outputs: VecDeque<f64>,
+}
+
+impl IIRFilter {
+    fn new(feedforward: Arc<Vec<f64>>, feedback: Arc<Vec<f64>>) -> Self {
+        Self {
+            feedforward,
+            feedback,
+            inputs: VecDeque::with_capacity(MAX_COEFFS),
+            outputs: VecDeque::with_capacity(MAX_COEFFS),
+        }
+    }
+
+    fn calculate_output(&mut self, input: f32) -> f32 {
+        self.inputs.push_front(input as f64);
+
+        if self.inputs.len() > MAX_COEFFS {
+            self.inputs.pop_back();
+        }
+
+        let inputs_sum = self
+            .feedforward
+            .iter()
+            .zip(self.inputs.iter())
+            .fold(0.0, |acc, (c, v)| acc + c * v);
+
+        let outputs_sum = self
+            .feedback
+            .iter()
+            .skip(1)
+            .zip(self.outputs.iter())
+            .fold(0.0, |acc, (c, v)| acc + c * v);
+
+        let output = (inputs_sum - outputs_sum) / self.feedback[0];
+
+        if output.is_nan() {
+            // Per spec:
+            // Note: The UA may produce a warning to notify the user that NaN values have occurred in the filter state.
+            // This is usually indicative of an unstable filter.
+            //
+            // But idk how to produce warnings
+            panic!("NaN in IIRFilter state");
+        }
+
+        self.outputs.push_front(output);
+
+        if self.outputs.len() > MAX_COEFFS {
+            self.outputs.pop_back();
+        }
+
+        output as f32
+    }
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct IIRFilterNode {
+    channel_info: ChannelInfo,
+    filters: Vec<IIRFilter>,
+    feedforward: Arc<Vec<f64>>,
+    feedback: Arc<Vec<f64>>,
+}
+
+impl IIRFilterNode {
+    pub fn new(options: IIRFilterNodeOptions, channel_info: ChannelInfo) -> Self {
+        if options.feedforward.len() == 0 {
+            panic!("NotSupportedError: feedforward must have at least one coeff");
+        }
+
+        if options.feedforward.len() > MAX_COEFFS {
+            panic!(
+                "NotSupportedError: feedforward max length is {}",
+                MAX_COEFFS
+            );
+        }
+
+        if options.feedforward.iter().all(|&v| v == 0.0_f64) {
+            panic!("InvalidStateError: all coeffs are zero");
+        }
+
+        if options.feedback.len() == 0 {
+            panic!("NotSupportedError: feedback must have at least one coeff");
+        }
+
+        if options.feedback.len() > MAX_COEFFS {
+            panic!("NotSupportedError: feedback max length is {}", MAX_COEFFS);
+        }
+
+        if options.feedback[0] == 0.0 {
+            panic!("InvalidStateError: first feedback coeff must not be zero");
+        }
```

Comment:
In theory we should never get to these `panic`s as these are DOM level checks. So we usually do not double check these conditions in servo-media. I think it would be better to add `debug_assert`s instead of `panic`s.